### PR TITLE
bash-completion: Add some bash completion to run command

### PR DESCRIPTION
**Pull request by sergio-costas**
The run command only offers bash completion for the parameters, but not for the snap application to run.

This patch adds some (incomplete) bash completion support for that, by using the folder list in /snap as source for completion when asking for a program. It is not perfect, because it's not possible to autocomplete other apps inside a snap, but is better than the current situation.

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

### CODE REVIEW

**Review by zyga — COMMENTED** [review]
Meta-comment, could we get completions from the go implementation of snap run directly and simply expose them like we do for other snap commands already? I think it might be less fragile than doing it all in shell.

**Comment by sergio-costas** [comment]
The only problem of this script is that, to get the runable apps inside the YAML file, it needs yq outside the snap, for BASH. Is there a way of doing that?

File: data/completion/bash/snap [hunk 1]
```diff
@@ -14,6 +14,33 @@
 #  You should have received a copy of the GNU General Public License
 #  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+_get_runable_snaps() {
+    snaplist=$(ls -1 -d /snap/${cur}* 2>/dev/null |cut -d / - -f 3-)
```
Comment (zyga) [inline data/completion/bash/snap:18]: This assumes `/snap`

**Review by zyga — COMMENTED** [review]
Meta-comment, could we get completions from the go implementation of snap run directly and simply expose them like we do for other snap commands already? I think it might be less fragile than doing it all in shell.

**Comment by sergio-costas** [comment]
Also, here the idea is to autocomplete first the name of the snap, and only if you delete the space at the end and add a dot, autocomplete the inner apps. Not sure if that is possible inside the GO code...

**Comment by sergio-costas** [comment]
@zyga Ok, I've been able to put everything inside Go code. Please, review it when yo have some spare time.

**Comment by sergio-costas** [comment]
Ok, it fails when trying to launch a "normal" command. Have to see why...

**Comment by zyga** [comment]
> Also, here the idea is to autocomplete first the name of the snap, and only if you delete the space at the end and add a dot, autocomplete the inner apps. Not sure if that is possible inside the GO code...

Of course it is. It's all just code.

**Comment by sergio-costas** [comment]
@zyga I know, I already did :-D Anyway, as commented in the chat, I'm not very convinced with the modifications to the tests, because now that the snap name is a positional parameter, it's not returned with ParseArgs(), and I'm not sure how to check that. Also, it is added in the help command, breaking the style...

**Comment by sergio-costas** [comment]
Ok, I think that it's ready for another review round.

File: cmd/snap/complete.go [hunk 2]
```diff
@@ -36,6 +37,95 @@ import (
 	"github.com/snapcore/snapd/snap"
 )
 
+// SnapAndApp holds a snap name and an application name
+type SnapAndApp struct {
+	Snap    string
+	App     string
+	has_dot bool
```
Comment (bboozzoo) [inline cmd/snap/complete.go:44]:
```suggestion
	hasDot bool
```

File: cmd/snap/complete.go [hunk 3]
```diff
@@ -36,6 +37,95 @@ import (
...
+			}
+			snapYamlFn := filepath.Join(snap.BaseDir(snap_t.Name), "current", "meta", "snap.yaml")
+			meta, err := os.ReadFile(snapYamlFn)
+			if err != nil {
+				continue
+			}
+			info, err := snap.InfoFromSnapYaml(meta)
+			if err != nil {
+				continue
+			}
```
Comment (bboozzoo) [inline cmd/snap/complete.go:101]: you can use snap.ReadCurrentInfo()

File: cmd/snap/complete.go [hunk 4]
```diff
@@ -36,6 +37,95 @@ import (
...
+	var matchSnap SnapAndApp
+	matchSnap.UnmarshalFlag(match)
+	snaps, err := mkClient().List(nil, nil)
+	if err != nil {
+		return nil
+	}
+	if !matchSnap.has_dot {
+		// No dot in match, so complete with snap names
+		ret := make([]flags.Completion, 0, len(snaps))
+		for _, snap_t := range snaps {
```
Comment (bboozzoo) [inline cmd/snap/complete.go:89]:
```suggestion
		for _, name := range snaps {
```

File: cmd/snap/complete.go [hunk 5]
```diff
@@ -36,6 +37,95 @@ import (
...
+	return nil
+}
+
+func (s SnapAndApp) Complete(match string) []flags.Completion {
+	var matchSnap SnapAndApp
+	matchSnap.UnmarshalFlag(match)
+	snaps, err := mkClient().List(nil, nil)
+	if err != nil {
+		return nil
+	}
```
Comment (bboozzoo) [inline cmd/snap/complete.go:85]: I think you only need this when completing a string that has no dot, in other case the user has already provided the snap name they want, so no need to call out to snapd API

File: cmd/snap/complete.go [hunk 6]
```diff
@@ -36,6 +37,95 @@ import (
...
+
+	snapYamlFn := filepath.Join(snap.BaseDir(matchSnap.Snap), "current", "meta", "snap.yaml")
+	meta, err := os.ReadFile(snapYamlFn)
+	if err != nil {
+		return nil
+	}
+	info, err := snap.InfoFromSnapYaml(meta)
+	if err != nil {
+		return nil
+	}
```
Comment (bboozzoo) [inline cmd/snap/complete.go:119]: snap.ReadCurrentInfo

File: cmd/snap/complete.go (hunk 3 above)
Comment (sergio-costas) [reply cmd/snap/complete.go]: Great suggestion! Done.

File: cmd/snap/complete.go (hunk 4 above)
Comment (sergio-costas) [reply cmd/snap/complete.go:88]: Not sure... that's not the snap name, but a `client.Snap` structure/object. The name is a property inside it...

File: cmd/snap/complete.go (hunk 5 above)
Comment (sergio-costas) [reply cmd/snap/complete.go:84]: Good point. Moved inside the `hasDot` if statement.

File: cmd/snap/complete.go (hunk 6 above)
Comment (sergio-costas) [reply cmd/snap/complete.go]: Done too.

**Comment by sergio-costas** [comment]
@bboozzoo Had to rebase because github couldn't rebase it automatically.

File: cmd/snap/complete.go (hunk 4 above)
Comment (sergio-costas) [reply cmd/snap/complete.go:88]: (and can't use `snap` as the name because it would conflict with the import...)

Comment (sergio-costas) [reply cmd/snap/complete.go]: I like it. Done.

File: cmd/snap/complete.go [hunk 7]
```diff
@@ -36,6 +36,84 @@ import (
...
+func (s SnapAndApp) Complete(match string) []flags.Completion {
+	var matchSnap SnapAndApp
+	matchSnap.UnmarshalFlag(match)
+	if !matchSnap.hasDot {
+		// No dot in match, so complete with snap names
+		installedSnaps, err := mkClient().List(nil, nil)
+		if err != nil {
+			return nil
+		}
+		ret := make([]flags.Completion, 0, len(installedSnaps))
```
Comment (bboozzoo) [inline cmd/snap/complete.go:87]: this could perhaps be too optimistic. Not all of the snaps can possibly match, maybe this could be tweaked to something like:
```suggestion
		// we're expecting only a subset of all snaps to match, so start
		// with a reasonable guess for capacity and allocate more as we
		// go through the list and find more matches
		const matchesCapacityGuess = 5
		ret := make([]flags.Completion, 0, matchesCapacityGuess)
```

I guess it doesn't matter that much since we already allocated the memory to hold a list of all installed snaps, but let's not allocate more if there isn't a clear need.

**Review by bboozzoo — APPROVED** [review]
LGTM

File: cmd/snap/complete.go (hunk 7 above)
Comment (sergio-costas) [reply cmd/snap/complete.go]: Changed to no reservation.

Comment (sergio-costas) [reply cmd/snap/complete.go]: It's not needed, because it's called only "sometimes", so the speed gain is negligible.

**Comment by sergio-costas** [comment]
Rebased against master.

**Comment by sergio-costas** [comment]
Rebased against master.

**Comment by bboozzoo** [comment]
Definitely needs a rebase. @sergio-costas can you? :)

**Comment by sergio-costas** [comment]
@bboozzoo Do you want me also to squash everything?

**Comment by sergio-costas** [comment]
squashed.

**Comment by sergio-costas** [comment]
Oh... something fails... Let me check...

**Comment by olivercalder** [comment]
@zyga we should look at this

File: cmd/snap/cmd_run.go [hunk 8]
```diff
@@ -282,15 +285,18 @@ func maybeCheckSystemKeyMismatch(cli *client.Client) (changeID string, err error
...
 func (x *cmdRun) Execute(args []string) error {
-	if len(args) == 0 {
-		return errors.New(i18n.G("need the application to run as argument"))
+	snapApp := x.Positionals.SnapName.FullName()
+	if len(snapApp) == 0 {
+		if len(args) == 0 {
+			return errors.New(i18n.G("need the application to run as argument 1"))
+		}
+		snapApp = args[0]
+		args = args[1:]
```
Comment (olivercalder) [inline cmd/snap/cmd_run.go:300]: When would we end up with non-empty `args[0]` holding the snap-app name but it not being parsed into `x.Positionals.SnapName`?

File: cmd/snap/main.go [hunk 9]
```diff
@@ -182,6 +182,11 @@ func lintArg(cmdName, optName, desc, origDesc string) {
 		// see comment in fixupArg about the >s case
 		return
 	}
+	if len(optName) > 0 && optName[0] == '<' && strings.HasSuffix(optName, "]") &&
+		strings.Contains(optName, ">") && strings.Contains(optName, "[") {
+		// see comment in fixupArg about the >s case
+		return
```
Comment (olivercalder) [inline cmd/snap/main.go:188]: This is fine but at some point a proper simple stack parser may be more robust. For example this would pass the linter but is clearly wrong:

```
<><][]
```

but probably the input arg is restricted prior to this anyway so it's probably fine

**Comment by olivercalder** [comment]
I'm rebasing to resolve conflict and get this ready to land.

File: cmd/snap/cmd_run.go (hunk 8 above)
Comment (sergio-costas) [reply cmd/snap/cmd_run.go:300]: Well, this is shocking... I removed that and it works. But I remember that I had to add it to make it pass the tests... maybe I did other changes after and that became not needed...

Comment (sergio-costas) [reply cmd/snap/cmd_run.go:300]: Do you want me to create a patch to remove it?

Comment (sergio-costas) [reply cmd/snap/cmd_run.go:300]: Ok, it seems that it's enough with

```
	snapApp := x.Positionals.SnapName.FullName()
	if len(snapApp) == 0 {
		return errors.New(i18n.G("need the application to run as argument 1"))
	}
```

Comment (sergio-costas) [reply cmd/snap/cmd_run.go:300]: Ok, no... the code is required. If I remove it and try `snapcraft clean` I receive this:
```
craft-providers error: Failed to list instances for project 'snapcraft'.
* Command that failed: 'lxc --project snapcraft list local: --format=yaml'
* Command exit code: 46
* Command standard error output: b'error interno, por favor rep\xc3\xb3rtelo: fall\xc3\xb3 al ejecutar "lxd.lxc" : race condition detected, snap-run can only retry once\n'
Full execution log: '/home/sergio.costas@canonical.com/.local/state/snapcraft/log/snapcraft-20251117-104533.284078.log'
```

Comment (sergio-costas) [reply cmd/snap/cmd_run.go:300]: Did some tests, and the problem is that when snap is called with `snap run snapcraft clean` everything works as expected, but when called with `snapcraft clean` the parser can't parse the data and is when this piece of code enters in action.

Comment (sergio-costas) [reply cmd/snap/cmd_run.go:300]: I can add a comment explaining it, if you want.

Comment (olivercalder) [reply cmd/snap/cmd_run.go:300]: A comment would be useful in a follow-up, thanks :) I mostly care that the code is actually correct. Better yet if this edge case is well-tested. Do you know if it could be tested in unit tests?

Comment (sergio-costas) [reply cmd/snap/cmd_run.go:300]: That a good question... I'll try to test it.

Comment (olivercalder) [reply cmd/snap/cmd_run.go:300]: Follow-up is here: https://github.com/canonical/snapd/pull/16267
